### ISSUE TITLE
wxWidgets fails to link

### ISSUE THREAD

**Issue by adrianbrink** [erlang/otp#7940]
**Describe the bug**
I'm installing erlang with `asdf install erlang 26.1`.

wxWidgets fails to link.

**To Reproduce**
`Darwin two 23.1.0 Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:45 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6020 arm64`

When I run `asdf install erlang 26.1` I receive this output
```
checking Checking wxWidgets for min version:... -mmacosx-version-min=14
checking if compiler accepts -ObjC... yes
checking if we can add -fno-move-loop-invariants to CXXNOOPT (via CFLAGS)... no
checking if we can add -fno-var-tracking-assignments to CXXNOOPT (via CFLAGS)... no
checking for GL/gl.h... no
checking for OpenGL/gl.h... yes
checking for GL/glu.h... no
checking for OpenGL/glu.h... yes
checking for wx-config... /opt/homebrew/bin/wx-config
checking for wxWidgets version >= 3.0.2 (--unicode)... yes (version 3.2.4)
checking for wxWidgets static library... no
checking for wxwidgets webview... yes
checking for wx-config... /opt/homebrew/bin/wx-config
checking for wxWidgets version >= 3.0.2 (--unicode --debug=yes)... yes (version 3.2.4)
checking for wxWidgets static library... no
checking for debug build of wxWidgets... yes
checking for wx-config... /opt/homebrew/bin/wx-config
checking for wxWidgets version >= 3.0.2 (--unicode --debug=no)... yes (version 3.2.4)
checking for wxWidgets static library... no
checking for standard build of wxWidgets... yes
checking for wxwidgets 3.0 compatibility ... yes
checking for wxwidgets opengl support... yes
checking for GLintptr... yes
checking for GLintptrARB... yes
checking for GLchar... yes
checking for GLcharARB... yes
checking for GLhalfARB... yes
checking for GLint64EXT... yes
checking GLU Callbacks uses Tiger Style... no
checking if we can link wxwidgets programs... no
checking if we can add -Werror=return-type to CFLAGS (via CFLAGS)... yes
checking if we can add -Werror=return-type to CXXFLAGS (via CFLAGS)... yes
configure: creating aarch64-apple-darwin23.1.0/config.status
config.status: creating config.mk
config.status: creating c_src/Makefile
configure: WARNING: Can not link wx program are all developer packages installed?

*********************************************************************
**********************  APPLICATIONS INFORMATION  *******************
*********************************************************************

wx             : Can not link wx program are all developer packages installed?

*********************************************************************
 MAKE   depend
 MAKE   generate
 GEN    aarch64-apple-darwin23.1.0/opt/jit/beam_asm_global.hpp
 GEN    aarch64-apple-darwin23.1.0/opt/jit/TABLES-GENERATED
 GEN    aarch64-apple-darwin23.1.0/opt/jit/OPCODES-GENERATED
 GEN    aarch64-apple-darwin23.1.0/erl_version.h
 GEN    aarch64-apple-darwin23.1.0/opt/jit/erl_alloc_types.h
 GEN    aarch64-apple-darwin23.1.0/opt/jit/driver_tab.c
 MAKE   opt
 GEN    obj/aarch64-apple-darwin23.1.0/opt/depend.mk
 GEN    aarch64-apple-darwin23.1.0/preload.c
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/simple_c_gc.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/re.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_lexer.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_main.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_node.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_string.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_parser.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_printers.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_yield_fun.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_symbol.o
 CC     /Users/two/.asdf/plugins/erlang/kerl-home/builds/asdf_26.1/otp_src_26.1/erts/lib_src/yielding_c_fun/aarch64-apple-darwin23.1.0/ycf_utils.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/r/ethr_aux.o
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-L/opt/homebrew/opt/unixodbc/lib' [-Wunused-command-line-argument]
 CC     obj/aarch64-apple-darwin23.1.0/opt/r/ethr_atomics.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/r/ethr_mutex.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/r/ethr_cbf.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/r/ethread.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/r/ethr_event.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/erl_printf_format.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/erl_printf.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/erl_misc_utils.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/r/erl_printf_format.o
 CC     obj/aarch64-apple-darwin23.1.0/opt/r/erl_printf.o
```

I'm setting the following lines in my `.zshrc` file
```
export KERL_CONFIGURE_OPTIONS="--with-odbc=/opt/homebrew/opt/unixodbc --with-ssl=/opt/homebrew/opt/openssl@3 --enable-wx --with-wx-config=/opt/homebrew/bin/wx-config"
export CPPFLAGS="${CPPFLAGS+"$CPPFLAGS "}-I/opt/homebrew/opt/unixodbc/include"
export LDFLAGS="${LDFLAGS+"$LDFLAGS "}-L/opt/homebrew/opt/unixodbc/lib"
export KERL_BUILD_DOCS=yes
export PATH="/opt/homebrew/opt/openjdk/bin:$PATH"
```

```
➜  ~ wx-config --version-full
3.2.4.0
```

**Expected behavior**
I would expect wxWidges to correctly link.

**Affected versions**
`26.1`

**Comment by cjschneider2** [comment]
I can reproduce this as well using Homebrew. Strangely though, compiling using the master branch from git (not using homebrew) with the required dependencies does work. I'm not done investigating on my side but it would seem like wx-config doesn't give a valid path to the libraries when using the homebrew build env.

**Comment by Sgoettschkes** [comment]
I am running into the same issue with Erlang 25.2. I tried figuring out what command the erlang installer uses to extract the path, but wasn't able to. Running `wx-config --list` locally with wxwidgets installed using homebrew returns the following:

```

    Default config is osx_cocoa-unicode-3.2

  Default config will be used for output

```

**Comment by taiansu** [comment]
Have the same issue when installing Erlang version from 26.1 to 26.2.1. All of them errored out with 
```
wxEventLoop* m_modalEventLoop = NULL;
```
in the log file.

**My env**
OS: macOS 14.2.1 on M1
XCode 15.1
wxwidgets: 3.2.4 (via homebrew with and without `--build-from-source` option)

**Comment by yarisx** [comment]
OTP failed to find wxWidgets on my Ubuntu 22.04 with wxWidgets 3.0.5 installed through apt. The lib/wx/configure script correctly determined that webview is available, but later(?) failed on "link" stage. After a short look into the configure script (and configure.ac) it seems to me that the script does not use paths to wxWidgets libs when linking.

**Comment by dgud** [comment]
It does pick up the patch from `wx-config` script.
You need to lock inside the config.log and see why it fails to link.

**Comment by yarisx** [comment]
```
diff --git a/lib/wx/configure.ac b/lib/wx/configure.ac
index d49399c..7e20c0f 100644
--- a/lib/wx/configure.ac
+++ b/lib/wx/configure.ac
@@ -666,6 +666,8 @@ saved_LIBS=$LIBS
 
 if test X"$WX_HAVE_STATIC_LIBS" = X"true" ; then
    LIBS=$WX_LIBS_STATIC   
+else
+   LIBS=$WX_LIBS
 fi
 
 AC_LINK_IFELSE([AC_LANG_SOURCE([[
```
Applying this patch and regenerating the corresponding `configure` fixed the issue for me on both systems.

**Comment by yarisx** [comment]
Part of the issue is fixed by #7670 , but only part. The build with `--enable-pie` is still broken on Linux, so I attach the patch that fixed both configuration and build issues on OSX and Linux for me. The patch incorporates the snippet from above (as does #7670), but it also filters out `-pie` `-fPIE` from supplied CFLAGS and LDFLAGS, since the resulting wx driver is a shared library. The patch also changes the type of the library for OSX from bundle to dylib, but I'm not sure that this change is 100% correct.
[wx_configure.ac.both_os.patch](https://github.com/erlang/otp/files/13884918/wx_configure.ac.both_os.patch)

**Comment by dgud** [comment]
If I check my config.log,  -LPATH is correct and added to config test. 
What am I missing?

@yarisx thanks for helping.

```
configure:6986: checking if we can link wxwidgets programs
configure:7016: g++ -o conftest  -Wall -fPIC -g -O2 -Wno-deprecated-declarations -fomit-frame-pointer -fno-strict-aliasing 
  -D_GNU_SOURCE -D_THREAD_SAFE -D_REENTRANT -DHAVE_GLIB=1 
  -I/home/uabdgud/src/wxWidgets/WX_3.2.4/lib/wx/include/gtk3-unicode-3.2 -I/home/uabdgud/src/wxWidgets/include
   -D_FILE_OFFSET_BITS=64 -DWXUSINGDLL -D__WXGTK__ -pthread  -D_GNU_SOURCE -D_THREAD_SAFE
    -D_REENTRANT -DHAVE_GLIB=1 -shared -fPIC  conftest.cpp
     -L/home/uabdgud/src/wxWidgets/WX_3.2.4/lib -pthread   -Wl,-rpath,/home/uabdgud/src/wxWidgets/WX_3.2.4/lib
      -lwx_gtk3u_stc-3.2 -lwx_gtk3u_xrc-3.2 -lwx_gtk3u_html-3.2 -lwx_baseu_xml-3.2 -lwx_gtk3u_core-3.2 -lwx_baseu-3.2
      -lwx_gtk3u_gl-3.2 -lwx_gtk3u_aui-3.2  >&5
```

**Comment by yarisx** [comment]
@dgud There is part of the issue illustrated by your quote: there is a `-shared` flag in the command. This makes the check useless: even if wx libraries are missing, building a shared library with references to missing functions will work:
```
~/tmp $ gcc -o test.so -shared test.c
gcc -o test.so -shared test.c
test.c: In function ‘main’:
test.c:4:15: warning: implicit declaration of function ‘call_a_missing_function’ [-Wimplicit-function-declaration]
    4 |     int res = call_a_missing_function();
      |               ^~~~~~~~~~~~~~~~~~~~~~~
~/tmp $ gcc -o test.so test.c
test.c: In function ‘main’:
test.c:4:15: warning: implicit declaration of function ‘call_a_missing_function’ [-Wimplicit-function-declaration]
    4 |     int res = call_a_missing_function();
      |               ^~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/ld: /tmp/cc6CDvWX.o: in function `main':
test.c:(.text+0x12): undefined reference to `call_a_missing_function'
collect2: error: ld returned 1 exit status
~/tmp $ cat test.c
#include <stdio.h>
int main(void)
{
    int res = call_a_missing_function();
    printf("Here I am: %d!", res);
    return 0;
}
```
The `-shared` flag is currently present in test commands on both maint-25 and master branches, so the tests pass even if wxWidgets installation is broken.

**Comment by yarisx** [comment]
It seems I created more chaos than order, apparently my branch that maps to maint-25 was not really up-to-date. Now time to make things clear.
So I tried to build wxWidgets support and these are issues I saw on Linux:
- wx libraries were not picked by configure when checking linking
- extra `-shared` flag used during link tests, rendering them somewhat useless
- when `--enable-pie` flag is used during configuration the subsequent build fails (although the flag's presence mitigates the previous issue by adding `-pie` to all test commands and thus causing tests to produce an executable)
The first issue is fixed by #7670 but the other two remain on master and maint-25:
- extra `-shared` flag is visible in your snippet
- the link command becomes this:
`gcc -shared -fPIC  -pie  x86_64-pc-linux-gnu/gl_nif.o x86_64-pc-linux-gnu/glu_nif.o x86_64-pc-linux-gnu/egl_impl.o -lGL -lGLU -o ../priv/x86_64-pc-linux-gnu/erl_gl.so` (see the `-pie` that does not belong) and the result looks similar to this:
`/usr/bin/ld: x86_64-pc-linux-gnu/gl_nif.o: undefined reference to symbol '_ZSt20__throw_length_errorPKc@@GLIBCXX_3.4'` (the exact undefined symbol may vary depending on Linux version and compiler)

On OSX bundling is not the same as creating a shared library, so the remaining two issues are not breaking the build (at least for me on Ventura/M1).
My patch attempts to fix all three issues, so it overlaps partly with the contents of #7670 (and contradicts w.r.t OSX link flags, but that can be removed from the patch).
Sorry for all the mess in my previous comments.
